### PR TITLE
Packet number bug fix

### PR DESCRIPTION
**Pull request by kevindweb**
Fixes a manager segmentation fault due to packet initialization.

## Summary:
When running multiple `speed_tester` NFs at once, shutting down and restarting, `rte_pktmbuf_alloc` sometimes cannot allocate all the packets. This would cause an issue with the number of packets in the following statement: 
```c
onvm_nflib_return_pkt_bulk(nf_info, pkts, pkts_generated);
``` 
The variable `pkts` had a length of `packet_number` (static uint32_t), but if we couldn't allocate the packet, we continue to the next try in the loop (`while` for LIBPCAP & `for` for `-c packet_num`). This pr fixes this by making a temporary counter to see how many packets were actually initialized. 

**Usage:**
Test by starting the manager with two NFs as such `./go.sh 1 -d 1 -c 16000` or `./go.sh 1 -d 1 -o pcap/pcap_file_here.pcap`. Then after both speed_testers are running on same destination, restart one going to another destination (`./go.sh 2 -d 2 -c 16000`). The manager should be able to handle this now. 

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                | 👍
| New functionality        |
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [x] PR is ready for review
 - [x] Test without the fix, and make sure the same process doesn't break with this PR 

## Review: 
- Sanity checks, assigned to @koolzz  @dennisafa 
  - Run make in /onvm and /examples directories
  - Code style, assigned to @koolzz  @dennisafa 

- Run linter
  - Check everything style related
  - Code design, assigned to @koolzz  @dennisafa 

- Check for memory leaks
  - Check that code is reusable
  - Code is clean, functions are concise
  - Verify that edge cases properly handled

- Performance, assigned to @koolzz  @dennisafa 
  - Run Speed Tester NF, report performance.

- Documentation, assigned to @koolzz  @dennisafa 
  - Check if the new changes are well documented, in both code and READMEs

### CODE REVIEW

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35132377

## Linter Passed

**Comment by koolzz** [comment]
Great find, I'll review later

File: examples/speed_tester/speed_tester.c [hunk 1]
```diff
@@ -424,16 +430,18 @@ nf_setup(struct onvm_nf_info *nf_info) {
                         onvm_ft_fill_key(&key, pkt);
                         pkt->hash.rss = onvm_softrss(&key);
 
-                        onvm_nflib_return_pkt(nf_info, pkt);
+                        /* Add packet to batch */
+                        pkts[i - 1] = pkt;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: This doesn't look pretty, can  you change it to `pkts[i] = pkt;`? The i++ increment should be after this statement

File: examples/speed_tester/speed_tester.c [hunk 2]
```diff
@@ -410,6 +413,9 @@ nf_setup(struct onvm_nf_info *nf_info) {
                         if (pkt == NULL)
                                 break;
 
+                        /* New packet generated successfully */
+                        pkts_generated++;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: Put this after we do `pkts[i] = pkt`, fits better there.

File: examples/speed_tester/speed_tester.c [hunk 3]
```diff
@@ -400,6 +401,8 @@ nf_setup(struct onvm_nf_info *nf_info) {
                 }
 
                 packet_number = (packet_number ? packet_number : MAX_PKT_NUM);
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: can you also change this to 
`packet_number = (use_custom_pkt_count ? packet_number : MAX_PKT_NUM);`

File: examples/speed_tester/speed_tester.c [hunk 4]
```diff
@@ -472,10 +483,11 @@ nf_setup(struct onvm_nf_info *nf_info) {
                         }
                         pkts[i] = pkt;
                 }
-                onvm_nflib_return_pkt_bulk(nf_info, pkts, packet_number);
+                onvm_nflib_return_pkt_bulk(nf_info, pkts, pkts_generated);
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: lets move the `onvm_nflib_return_pkt_bulk` function out of the ifdef statement as its the same for both

Comment (kevindweb) [reply examples/speed_tester/speed_tester.c]: When I do this, it seems to think that pkts is declared but never used, any suggestions? The other fixes seem to be fine

Comment (koolzz) [reply examples/speed_tester/speed_tester.c]: If its being weird because of our #ifdef structure that its okay, we can keep it

Comment (kevindweb) [reply examples/speed_tester/speed_tester.c]: I think we can just take out the #ifdef LIBPCAP and keep the `pcap_filename != NULL`, then in the initialization of pkts, test whether the packet_number is changed? I can test and push to see your thoughts

Comment (koolzz) [reply examples/speed_tester/speed_tester.c]: The #ifdef LIBPCAP allows us to build without a pcap library

Comment (kevindweb) [reply examples/speed_tester/speed_tester.c]: Oh ok that makes sense

File: examples/speed_tester/speed_tester.c [hunk 5]
```diff
@@ -375,6 +375,7 @@ run_advanced_rings(struct onvm_nf_info *nf_info) {
 void
 nf_setup(struct onvm_nf_info *nf_info) {
         uint32_t i;
+        uint32_t pkts_generated = 0;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: Style nit, don't define variables at declaration,

File: examples/speed_tester/speed_tester.c [hunk 6]
```diff
@@ -424,16 +427,20 @@ nf_setup(struct onvm_nf_info *nf_info) {
                         onvm_ft_fill_key(&key, pkt);
                         pkt->hash.rss = onvm_softrss(&key);
 
-                        onvm_nflib_return_pkt(nf_info, pkt);
+                        /* Add packet to batch, and update counter */
+                        pkts[i] = pkt;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: Kind of feeling a `pkts[i++] = pkt;`, would look nicer

File: examples/speed_tester/speed_tester.c [hunk 7]
```diff
@@ -445,6 +452,9 @@ nf_setup(struct onvm_nf_info *nf_info) {
                                 break;
                         }
 
+                        /*new packet generated successfully*/
+                        pkts_generated++;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: Yeah move down to pkts[i] = pkt

File: examples/speed_tester/speed_tester.c [hunk 8]
```diff
@@ -399,10 +400,13 @@ nf_setup(struct onvm_nf_info *nf_info) {
                         rte_exit(EXIT_FAILURE, "Cannot open pcap file\n");
                 }
 
-                packet_number = (packet_number ? packet_number : MAX_PKT_NUM);
+                packet_number = (use_custom_pkt_count ? packet_number : MAX_PKT_NUM);
+                struct rte_mbuf *pkts[packet_number];
+
                 i = 0;
+                pkts_generated = 0;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: `pkts_generated = 0` should be before #ifdef LIBPCAP

**Comment by koolzz** [comment]
@onvm how does this look

**Comment by onvm** [comment]
> @onvm how does this look

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm how does this look

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35216053

## Linter Passed

File: examples/speed_tester/speed_tester.c [hunk 9]
```diff
@@ -470,12 +477,16 @@ nf_setup(struct onvm_nf_info *nf_info) {
...
+                        /* New packet generated successfully */
                         pkts[i] = pkt;
+                        pkts_generated++;
                 }
-                onvm_nflib_return_pkt_bulk(nf_info, pkts, packet_number);
+                onvm_nflib_return_pkt_bulk(nf_info, pkts, pkts_generated);
 #ifdef LIBPCAP
         }
 #endif
+        packet_number = pkts_generated;
```
Comment (koolzz) [inline examples/speed_tester/speed_tester.c]: if packet_number==0
   //no packets, print error msg, cry, exit

Comment (koolzz) [reply examples/speed_tester/speed_tester.c]: Actually, check if we wanted to generate 0 packets. If we did than we're fine, but If we wanted 128 packets and got 0 we should just error out and exit

**Comment by koolzz** [comment]
Fixes the bug, code looks good & works as expected, implement my last comment and I'll merge

**Comment by koolzz** [comment]
@onvm

**Comment by onvm** [comment]
> @onvm 

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm 

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35152533

## Linter Failed

examples/speed_tester/speed_tester.c:490:  Missing space before ( in if(  [whitespace/parens] [5]
Total errors found: 1

**Comment by koolzz** [comment]
@onvm style nit fixed, how about now

**Comment by onvm** [comment]
> @onvm style nit fixed, how about now

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm style nit fixed, how about now

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35173046

## Linter Passed

**Pull request merged**